### PR TITLE
Run ILLIXR headlessly

### PR DESCRIPTION
**Pull request by mhuzai**
We would like to be able to run without a physical display. This is useful during SSH connections and useful for profiling.

### CODE REVIEW

File: timewarp_gl/shaders/basic_shader.hpp [hunk 1]
```diff
@@ -1,6 +1,6 @@
 #include <GL/gl.h>
 
-#define GLSL_VERSION            "430 core"
+#define GLSL_VERSION            "330 core"
```
[the remark this answers is not in the document]
Comment (mhuzai) [reply timewarp_gl/shaders/basic_shader.hpp:3]: Nice idea. I can't decide which option is better. Maybe 2 because 1 would require a new file that just has one line in it?

[the remark this answers is not in the document]
Comment (charmoniumQ) [reply timewarp_gl/shaders/basic_shader.hpp]: I've implemented 2.

**Comment by charmoniumQ** [comment]
I'm approving this PR. However, I recommend two future changes:

1. After #112 is merged, use the external tool loader, instead of a dedicated loader.
2. Run headless in the CI pipeline. This will inform us about potentially-introduced runtime errors (i.e. segfaults).

**Review by charmoniumQ — COMMENTED** [review]
I've only made changes to the CI, and those are minor, so I am approving this.

File: timewarp_gl/shaders/timewarp_shader.hpp [hunk 2]
```diff
@@ -1,7 +1,5 @@
 #include <GL/gl.h>
 
-#define GLSL_VERSION            "430 core"
-
 const char* const timeWarpChromaticVertexProgramGLSL =
 	"#version " GLSL_VERSION "\n"
```
Comment (Zee2) [inline timewarp_gl/shaders/timewarp_shader.hpp:4]: `GLSL_VERSION` macro is still being used in the shader source despite the `#define` being removed above?

Comment (charmoniumQ) [reply timewarp_gl/shaders/timewarp_shader.hpp:4]: It's being defined at the command-line invocation of GCC. This way, it doesn't have to be duplicated in every shader.

Comment (mhuzai) [reply timewarp_gl/shaders/timewarp_shader.hpp:4]: After seeing Finn's comment I'm wondering if putting the GLSL version in the makefile is the most readable choice. Perhaps it should be in an include of its own.

File: docs/getting_started.md [hunk 3]
```diff
@@ -48,20 +52,11 @@ of applications.  Monado only supports Ubuntu 18.10, because of a low-level driv
...
-    path: ../monado_integration
-  openxr_app:
-    path: ../Monado_OpenXR_Simple_Example
-```
+5. Inspect `configs/monado.yaml`. The schema definition (with documentation inline) is in `runner/config_schema.yaml`.
 
-5. Compile and run:
+6. Compile and run:
 
-        ./runner.sh config.yaml
```
Comment (charmoniumQ) [inline docs/getting_started.md:64]: :+1: for fixing out-of-date documentation

File: timewarp_gl/shaders/timewarp_shader.hpp (hunk 2 above)
Comment (charmoniumQ) [reply timewarp_gl/shaders/timewarp_shader.hpp:4]: If you do, put it in `common/`.

File: common/extended_window.hpp [hunk 4]
```diff
@@ -43,89 +48,105 @@ namespace ILLIXR{
...
-            win = XCreateWindow(dpy, root, 0, 0, width, height, 0, vi->depth, InputOutput, vi->visual, CWColormap | CWEventMask, &swa);
-            XMapWindow(dpy, win);
+
+#ifndef NDEBUG
+            printf("Creating window\n");
+#endif
+            XSetWindowAttributes attributes;
+            attributes.colormap = cmap;
+            attributes.background_pixel = 0;
+            attributes.border_pixel = 0;
```
Comment (charmoniumQ) [inline common/extended_window.hpp:110]: Are these changes necessary for headless or something else?

File: common/extended_window.hpp [hunk 5]
```diff
@@ -43,89 +48,105 @@ namespace ILLIXR{
...
-            int context_attribs[] =
-                {
-                    GLX_CONTEXT_MAJOR_VERSION_ARB, 4,
-                    GLX_CONTEXT_MINOR_VERSION_ARB, 0,
-                    //GLX_CONTEXT_FLAGS_ARB        , GLX_CONTEXT_FORWARD_COMPATIBLE_BIT_ARB,
-                    None
-                };
+                glXGetProcAddressARB((const GLubyte *) "glXCreateContextAttribsARB");
+            int context_attribs[] = {
+                GLX_CONTEXT_MAJOR_VERSION_ARB, 3,
```
Comment (charmoniumQ) [inline common/extended_window.hpp:131]: Maybe this should also be defined for the project in an include file.

File: common/extended_window.hpp (hunk 4 above)
Comment (mhuzai) [reply common/extended_window.hpp:110]: This is _the_ fix for headless.

**Review by apache8080 — APPROVED** [review]
LGTM
